### PR TITLE
feat(api): wait for Init status after submitProgram

### LINKED ISSUE #3: Wait for Init status after submitProgram
### Problem to Solve

similar to #2 

### Possible Solution

submit the `submit_program` and listen to events, exit after program initialized

### Notes

_No response_

### PR DESCRIPTION
**Pull request by clearloop**
Resolves #3 .

- [x] upgrade runtime-spec to 960
- [x] wait till `InBlock`
- [x] add event api
- [x] filter init status ( InitSuccess / InitFailure ) 

@reviewer-or-team

## Example output

if success 

```
 4:44PM ./target/release/gear deploy ../ping/target/wasm32-unknown-unknown/release/demo_ping.wasm
Submited extrinsic Gear::submit_program
        Status: Ready
        Status: Broadcast( ["12D3KooWRv6hAyWTjs6CaJkwYGdG7YbJf8ZtarxBnMWNnywf9UHY", "12D3KooWJtDTQTX58AjNbCK4AHrpf1uXzTRyNYQoahcuUYWDcVZ6", "12D3KooWSZXSxiLBoKBQFHjrdg7DV4Wjf3Qdhea7h2FQhxPBiCEs", "12D3KooWGe2JaDzPhi9eik5ttTz1igKuDFij5sZzFPHFaRodkh9C", "12D3KooWBMAPjBTUxheMeWyCrxYn1kKX3pGBh3NCDnBiDj43EjLN", "12D3KooWDFf1g2LghBjcH6DMb5y1maCj1BqYt98VNh89znLSnCyu", "12D3KooWRGfgDJMYJm6zZeQyhum1gdCjQNpx2KXiK7MMF68Q4dZY", "12D3KooWFJsjWDBPgNSrsrVrNBBXmdrKBHXShq4eMKRKW8ZmAtb3", "12D3KooWKUBmBxceECmWEpdKrZ74bfBGFT19m5aG32nLT5D62dpL", "12D3KooWAri3f5JAgPWT7zSwSP5qXofJbX1CtY8QCcAm51aUVmb9", "12D3KooWPo4TeDiu4X13utPtsgHdvD1kHwaZ7CZMvuXhwpLpPigx", "12D3KooWNxqwAhwdCpFZQcspSyifqpoouhrmpE8kBxfeYQyZTefr", "12D3KooWD5y4kR8E7YNFVBrPyTM15TFR8SpaET4r2Fa1uzi2xkAn", "12D3KooWG5G3jaQ8ZVMCSTbyTRjfzhSpn55f8DGKDpFKvU4A3SPQ", "12D3KooWAaZjbQPrbjDgD2mkWFMXgot5Q3dHAG8G94Qosajt2U9M", "12D3KooWJptbqKsSphtwTZNK85N6TERme2KRGSx33W7KCCGYfjnJ", "12D3KooWSJ9DwKHZAZJqz9F4dYMokPfiThpHyHEQf11vzhggLq3o", "12D3KooWCfm5T1uRq3rCNzqjsq3eFwu8fjHkUo1XBZmZvo611BwA", "12D3KooWCWLK9UJsCsuayToWacd9ys7Kx4tbQxYVeLAdKRFxkoaH", "12D3KooWNwq32KceEgx1iRBsFseETUveDJ3W7SgAQRyWsbEeiT3Z", "12D3KooWPLQncdCUDdYJ4Zqi3uwRLYmgHc9gEpzryoE2LDswr2Fb", "12D3KooWPUGdJH6T9HvkMbZfcqeayNF9fKptJfm53hci9cnQkyfs", "12D3KooWSi26r5MwfVt2Zhsv5j44JQoeej9SCcPoiygKNTKH9X1q", "12D3KooWGNR4CeT1F514UWY7CSizqHXmQYkrAhC4dHD246G8go8e", "12D3KooWCpLwWWiLPV1CirjVNQ77S9Bz4nUPSz6GgPCZhg7LYSzN", "12D3KooWST7AyiRQgXcws8FbMEhcBDxKxnaHWVu5TEM5hmgsTGuJ", "12D3KooWHsTYMFNk7Cq5VVy63FWbsFHq2tZ9vPiDayeDhh4kqrSN", "12D3KooWCu3VahHx8VFvExcjNUX76zLLaCYvMoHrvqhdUSuXKFS7", "12D3KooWKAz1fDTWFG2uSrKxGJxAnawWSQYRPL5magzHejiMDBc3", "12D3KooWJdvuMq2WfsAP22kkoKSksgXU95WB6uE5KPT4nQu2V3e9", "12D3KooWCfh1fb16y3q2uE2JmkuRM4Qv7L9gpmuXvQhxun1AJLBt", "12D3KooWR2Cx9nKyHGn3hiW51eLhkeVma5iTTPePRcqftDzvDj9d", "12D3KooWHu9argxVpL7aFL31fafX7HXZCsRy4HaoHU2jbK3jfYju", "12D3KooWGBVkyuDarCR8F92AAKD7dDSoK8VmkBBynyDViG3QAY1H", "12D3KooWAUeuxMCrm8gmxbGSxqcYow3YJ7NrAj3gDuZk55vYg419", "12D3KooWAXVLcPgnNtZpijkwadGkvyAdKirzPnhUXbvhrwXYR8JR"] )
        Status: InBlock( block_hash: 0x109d…bdb9, extrinsic_hash: 0x5183…fd9e )
        CodeSaved(0x8195b7678b92217e299124d1d0b704f546af93fb352d896e053081102d9fba38)
        InitMessageEnqueued(MessageInfo { message_id: MessageId([69, 156, 255, 213, 58, 40, 237, 67, 46, 108, 121, 218, 112, 237, 155, 24, 65, 79, 125, 9, 82, 60, 95, 196, 189, 248, 27, 12, 169, 177, 175, 181]), program_id: ProgramId([84, 153, 21, 210, 20, 252, 14, 16, 132, 193, 247, 116, 21, 25, 193, 134, 251, 101, 180, 131, 100, 228, 10, 114, 96, 124, 184, 252, 113, 198, 111, 11]), origin: 0x32420a30ee192840a65abe4416cd129f9160f57dab84f6612a6973d2891fe069 })
        InitSuccess(MessageInfo { message_id: MessageId([69, 156, 255, 213, 58, 40, 237, 67, 46, 108, 121, 218, 112, 237, 155, 24, 65, 79, 125, 9, 82, 60, 95, 196, 189, 248, 27, 12, 169, 177, 175, 181]), program_id: ProgramId([84, 153, 21, 210, 20, 252, 14, 16, 132, 193, 247, 116, 21, 25, 193, 134, 251, 101, 180, 131, 100, 228, 10, 114, 96, 124, 184, 252, 113, 198, 111, 11]), origin: 0x32420a30ee192840a65abe4416cd129f9160f57dab84f6612a6973d2891fe069 })
        Status: Finalized( block_hash: 0x109d…bdb9, extrinsic_hash: 0x5183…fd9e )
Successfully submited call Gear::submit_program 0x5183…fd9e at 0x109d…bdb9!
```

if failed

```
 4:46PM ./target/release/gear deploy ../ping/target/wasm32-unknown-unknown/release/demo_ping.wasm
Submited extrinsic Gear::submit_program
        Status: Ready
        Status: Broadcast( ["12D3KooWRv6hAyWTjs6CaJkwYGdG7YbJf8ZtarxBnMWNnywf9UHY", "12D3KooWJtDTQTX58AjNbCK4AHrpf1uXzTRyNYQoahcuUYWDcVZ6", "12D3KooWSZXSxiLBoKBQFHjrdg7DV4Wjf3Qdhea7h2FQhxPBiCEs", "12D3KooWGe2JaDzPhi9eik5ttTz1igKuDFij5sZzFPHFaRodkh9C", "12D3KooWBMAPjBTUxheMeWyCrxYn1kKX3pGBh3NCDnBiDj43EjLN", "12D3KooWDFf1g2LghBjcH6DMb5y1maCj1BqYt98VNh89znLSnCyu", "12D3KooWRGfgDJMYJm6zZeQyhum1gdCjQNpx2KXiK7MMF68Q4dZY", "12D3KooWFJsjWDBPgNSrsrVrNBBXmdrKBHXShq4eMKRKW8ZmAtb3", "12D3KooWKUBmBxceECmWEpdKrZ74bfBGFT19m5aG32nLT5D62dpL", "12D3KooWAri3f5JAgPWT7zSwSP5qXofJbX1CtY8QCcAm51aUVmb9", "12D3KooWPo4TeDiu4X13utPtsgHdvD1kHwaZ7CZMvuXhwpLpPigx", "12D3KooWNxqwAhwdCpFZQcspSyifqpoouhrmpE8kBxfeYQyZTefr", "12D3KooWD5y4kR8E7YNFVBrPyTM15TFR8SpaET4r2Fa1uzi2xkAn", "12D3KooWG5G3jaQ8ZVMCSTbyTRjfzhSpn55f8DGKDpFKvU4A3SPQ", "12D3KooWAaZjbQPrbjDgD2mkWFMXgot5Q3dHAG8G94Qosajt2U9M", "12D3KooWJptbqKsSphtwTZNK85N6TERme2KRGSx33W7KCCGYfjnJ", "12D3KooWSJ9DwKHZAZJqz9F4dYMokPfiThpHyHEQf11vzhggLq3o", "12D3KooWCfm5T1uRq3rCNzqjsq3eFwu8fjHkUo1XBZmZvo611BwA", "12D3KooWCWLK9UJsCsuayToWacd9ys7Kx4tbQxYVeLAdKRFxkoaH", "12D3KooWNwq32KceEgx1iRBsFseETUveDJ3W7SgAQRyWsbEeiT3Z", "12D3KooWPLQncdCUDdYJ4Zqi3uwRLYmgHc9gEpzryoE2LDswr2Fb", "12D3KooWPUGdJH6T9HvkMbZfcqeayNF9fKptJfm53hci9cnQkyfs", "12D3KooWSi26r5MwfVt2Zhsv5j44JQoeej9SCcPoiygKNTKH9X1q", "12D3KooWGNR4CeT1F514UWY7CSizqHXmQYkrAhC4dHD246G8go8e", "12D3KooWCpLwWWiLPV1CirjVNQ77S9Bz4nUPSz6GgPCZhg7LYSzN", "12D3KooWST7AyiRQgXcws8FbMEhcBDxKxnaHWVu5TEM5hmgsTGuJ", "12D3KooWHsTYMFNk7Cq5VVy63FWbsFHq2tZ9vPiDayeDhh4kqrSN", "12D3KooWCu3VahHx8VFvExcjNUX76zLLaCYvMoHrvqhdUSuXKFS7", "12D3KooWKAz1fDTWFG2uSrKxGJxAnawWSQYRPL5magzHejiMDBc3", "12D3KooWJdvuMq2WfsAP22kkoKSksgXU95WB6uE5KPT4nQu2V3e9", "12D3KooWCfh1fb16y3q2uE2JmkuRM4Qv7L9gpmuXvQhxun1AJLBt", "12D3KooWR2Cx9nKyHGn3hiW51eLhkeVma5iTTPePRcqftDzvDj9d", "12D3KooWHu9argxVpL7aFL31fafX7HXZCsRy4HaoHU2jbK3jfYju", "12D3KooWGBVkyuDarCR8F92AAKD7dDSoK8VmkBBynyDViG3QAY1H", "12D3KooWAUeuxMCrm8gmxbGSxqcYow3YJ7NrAj3gDuZk55vYg419", "12D3KooWAXVLcPgnNtZpijkwadGkvyAdKirzPnhUXbvhrwXYR8JR"] )
        Status: InBlock( block_hash: 0x645c…c270, extrinsic_hash: 0xa444…32e6 )
        Status: Finalized( block_hash: 0x645c…c270, extrinsic_hash: 0xa444…32e6 )
Successfully submited call Gear::submit_program 0xa444…32e6 at 0x645c…c270!
Module error: Gear: ProgramAlreadyExists

Program already exists.

Occurs if a program with some specific program id already exists in program storage.
```

### CODE REVIEW

File: src/cmd/deploy.rs [hunk 1]
```diff
@@ -63,8 +81,39 @@ impl Deploy {
...
     }
+
+    async fn wait_init_status(&self, mut events: Events<'_>) -> Result<()> {
+        while let Some(events) = events.next().await {
+            for maybe_event in events?.iter() {
+                let event = maybe_event?.event;
+
+                // Exit when extrinsic failed.
+                //
+                // # Safty
```
Comment (shamilsan) [inline src/cmd/deploy.rs:98]: Typo: 
```suggestion
                // # Safety
```

File: src/cmd/deploy.rs [hunk 2]
```diff
@@ -63,8 +81,39 @@ impl Deploy {
...
+                //
+                // The error message will be paniced in another thread.
```
Comment (shamilsan) [inline src/cmd/deploy.rs:100]: Typo:
```suggestion
                // The error message will be panicked in another thread.
```

**Review by shamilsan — APPROVED** [review]
Awesome ❤️ 

But I'd recommend adding some smoke tests to CI against the latest node running in the dev node (it can be downloaded here: https://github.com/gear-tech/gear/releases/download/build/gear-node).

We run the node in a detached mode for such kind of purpose: https://github.com/gear-tech/gear/blob/master/scripts/src/test.sh#L98

**Comment by clearloop** [comment]
> But I'd recommend adding some smoke tests to CI against the latest node running in the dev node (it can be downloaded here: https://github.com/gear-tech/gear/releases/download/build/gear-node).

but some functions/events only work on the matched spec_version, if we align to the lastest version of gear-node, this program may not work for the gear testnet

**Comment by shamilsan** [comment]
> but some functions/events only work on the matched spec_version, if we align to the lastest version of gear-node, this program may not work for the gear testnet

We are to keep the API/runtime part up to date and detect incompatibility with the node at early stages (another task to CI I guess, maybe some nighly job that compares actual API with the `generated.rs`).

Ideally, every new feature should be accompanied by the corresponding test/example.

**Comment by clearloop** [comment]
> > but some functions/events only work on the matched spec_version, if we align to the lastest version of gear-node, this program may not work for the gear testnet
> 
> We are to keep the API/runtime part up to date and detect incompatibility with the node at early stages (another task to CI I guess, maybe some nighly job that compares actual API with the `generated.rs`).
> 
> Ideally, every new feature should be accompanied by the corresponding test/example.

Agree to test if `gear-program` is compatible with gear node, but I'm considering should we align to the latest gear-node on https://github.com/gear-tech/gear#master or the testnet,

- if https://github.com/gear-tech/gear#master, we can download the binary and check
- if testnet, we can just query the rpc to check

**Comment by shamilsan** [comment]
> Agree to test if `gear-program` is compatible with gear node, but I'm considering should we align to the latest gear-node on https://github.com/gear-tech/gear#master or the testnet,
> 
>     * if https://github.com/gear-tech/gear#master, we can download the binary and check
> 
>     * if testnet, we can just query the rpc to check

Good question, I hope they will correspond one to another in the future (we will make runtime upgrades more often or strongly use semantic versioning).
I'd test it with the latest node binary at the moment. Or even with both to detect incompatibilities and execute the runtime upgrade. But it is up to you. Any test will be better than nothing.

**Pull request merged**